src/D2L.CodeStyle.Analyzers/Async/BlockingAnalyzer.cs

```diff
@@ -0,0 +1,234 @@
+					noteMethodWithBlockingAttribute( methodSymbol );
+				}
+
+				return;
+			}
+
+			var attrData = GetAttribute( methodSymbol, blockingAttr );
+
+			if( attrData == null ) {
+				return;
+			}
+
+			ctx.ReportDiagnostic(
+				Diagnostic.Create(
+					Diagnostics.AsyncMethodCannotBeBlocking,
+					attrData.ApplicationSyntaxReference.GetSyntax().GetLocation(),
+					methodSymbol.Name
+				)
+			);
+
+			// TODO: code fix to remove [Blocking]
+		}
+
+		private static void InspectMethodCall(
+			SyntaxNodeAnalysisContext ctx,
+			INamedTypeSymbol blockingAttr,
+			Action<IMethodSymbol> noteMethodHasBlockingCall
+		) {
+			// Goal: find calls to blocking methods inside non-blocking methods, or in properties.
+
+			var invocation = (InvocationExpressionSyntax)ctx.Node;
+
+			var invokedSymbolInfo = ctx.SemanticModel.GetSymbolInfo( invocation.Expression );
+
+			// Unfortunately there are ways to use this to dodge analysis, but
+			// there is only so much we can do.
+			if( invokedSymbolInfo.Symbol?.Kind != SymbolKind.Method ) {
+				return;
+			}
+
+			var invokedMethodSymbol = (IMethodSymbol)invokedSymbolInfo.Symbol;
+
+			// async methods are never [Blocking] (see other parts of the analyzer)
+			if( invokedMethodSymbol.IsAsync ) {
+				return;
+			}
+
+			// If the thing we're calling isn't [Blocking] then OK.
+			// TODO: support other means of inferring blockingness, for 3rd party libraries.
+			if( !HasAttribute( invokedMethodSymbol, blockingAttr ) ) {
+				return;
+			}
+
+			var myMethodDeclaration = invocation.Ancestors()
+				.FirstOrDefault( s => s.IsKind( SyntaxKind.MethodDeclaration ) )
+				as MethodDeclarationSyntax;
+
+			if( myMethodDeclaration == null ) {
+				ctx.ReportDiagnostic(
+					Diagnostic.Create(
+						Diagnostics.OnlyCallBlockingMethodsFromMethods,
+						invocation.GetLocation(),
+						invokedMethodSymbol.Name
+					)
+				);
+
+				return;
+			}
+
+			// Need to go through the semantic model, because overrides inherit [Blocking]
+			var myMethodSymbol = ctx.SemanticModel.GetDeclaredSymbol( myMethodDeclaration );
+
+			if( myMethodSymbol == null ) {
+				return;
+			}
+
+			if( HasAttribute( myMethodSymbol, blockingAttr ) ) {
+				// We have [Blocking] so we're allowed to call other [Blocking] things,
+				// but make note that we definitely require blocking.
+				noteMethodHasBlockingCall( myMethodSymbol );
+				return;
+			}
+
+			if( myMethodSymbol.IsAsync ) {
+				ctx.ReportDiagnostic(
+					Diagnostic.Create(
+						Diagnostics.AsyncMethodCannotCallBlockingMethod,
+						invocation.GetLocation(),
+						invokedMethodSymbol.Name
+					)
+				);
+
+				// TODO: code fix that looks for an async equivalent and suggests calling it instead.
+				return;
+			}
+
+			ctx.ReportDiagnostic(
+				Diagnostic.Create(
+					Diagnostics.BlockingCallersMustBeBlocking,
+					myMethodDeclaration.Identifier.GetLocation(),
+					invokedMethodSymbol.Name,
+					myMethodSymbol.Name
+				)
+			);

```

Comment:
Would this be more appropriately placed on the invocation (same as the async-cannot-call)?

Reply:
Hm, seems reasonable. I guess I was thinking that is (approximately) where the suggested code fix goes, but the relationship doesn't need to be so tight.

Reply:
Done